donkeycar/templates/donkey2.py

```diff
@@ -132,48 +136,102 @@ def drive_mode(mode,
 
 
 
+def load_image(path):
+    img = Image.open(path)
+    return np.array(img)
+
+def get_generator(input_keys, output_keys, record_paths, meta, tub_path):
+    while True:
+        for record_path in record_paths:
+            with open(record_path, 'r') as record_file:
+                record = json.load(record_file)
+                inputs = [record[key] for key in input_keys]
+                outputs = [record[key] for key in output_keys]
+                input_types = [meta[key] for key in input_keys]
+                # output_types = [meta[key] for key in output_keys]
+                for i in range(len(inputs)):
+                    type = input_types[i]
+                    if (type == 'image_array'):
+                        inputs[i] = load_image("%s/%s" % (tub_path, inputs[i]))
+                    elif (type == 'custom/prev_image'):
+                        # Currently previous images are in array, but there is only one
+                        imagePath = inputs[i][0]
+                        inputs[i] = load_image("%s/%s" % (tub_path, imagePath))
+                yield inputs, outputs
+
+def get_batch_generator(input_keys, output_keys, records, meta, tub_path):
+    # Yield here a tuple (inputs, outputs)
+    # both having arrays with batch_size length like:
+    # 0: [input_1[batch_size],input_2[batch_size]]
+    # 1: [output_1[batch_size],output_2[batch_size]]
+    record_gen = get_generator(input_keys, output_keys, records, meta, tub_path)
+    while True:
+        raw_batch = [next(record_gen) for _ in range(cfg.BATCH_SIZE)]
+        inputs = [[] for _ in range(len(input_keys))]
+        outputs = [[] for _ in range(len(output_keys))]
+        for rec in raw_batch:
+            for i in range(len(input_keys)):
+                inputs[i].append(rec[0][i])
+            for i in range(len(output_keys)):
+                outputs[i].append(rec[1][i])
+        numpyInputs = [np.asarray(ar) for ar in inputs]
+        numpyOutputs = [np.asarray(ar) for ar in outputs]
+        yield numpyInputs, numpyOutputs
+
+def get_meta(path):
+    with open('%s/meta.json' % path, 'r') as f:
+        meta = json.load(f)
+        meta_dict = {}
+        for i, key in enumerate(meta['inputs']):
+            meta_dict[key] = meta['types'][i]
+        return meta_dict
+            # TODO: filter out values not listed in inputs or outputs
+
+
+def get_train_val_gen(inputs, outputs, tub_names):
+    print('Loading data', tub_names)
+    print('Inputs', inputs)
```

Comment:
Debug code left here?

Can the tub_names, inputs or outputs be really long? Meaning is there a change that the std.out will get flooded.

Reply:
Maybe use logging module https://docs.python.org/3/library/logging.html?

Reply:
These could be interesting when changing the implementation of the neural network so makes sense to log when started